Authorization/Authorization/Presentation/Login/SignInView.swift

```diff
@@ -16,137 +16,170 @@ public struct SignInView: View {
     @ObservedObject
     private var viewModel: SignInViewModel
     
-    public init(viewModel: SignInViewModel) {
+    private var navigationController: UINavigationController
+    
+    public init(viewModel: SignInViewModel, navigationController: UINavigationController) {
         self.viewModel = viewModel
+        self.navigationController = navigationController
+    }
+
+    public func webLogin() async {
+        await viewModel.login(viewController: self.navigationController)
     }
     
     public var body: some View {
-        ZStack(alignment: .top) {
-            VStack {
-                CoreAssets.authBackground.swiftUIImage
-                    .resizable()
-                    .edgesIgnoringSafeArea(.top)
-            }.frame(maxWidth: .infinity, maxHeight: 200)
-            
-            VStack(alignment: .center) {
-                CoreAssets.appLogo.swiftUIImage
-                    .resizable()
-                    .frame(maxWidth: 189, maxHeight: 54)
-                    .padding(.vertical, 40)
+        if viewModel.forceWebLogin {
+            // On first load, we should bring the user right to the web login
+            // interface rather than showing this view.
```

Comment:
What do you think about this: instead of this boolean `forceWebLogin` which special-cases one login method, could we have a setting called `presumeLoginMethod` or `forceLoginMethod` which can optionally be set to one of the `LoginMethod` enum values?

Whether you call it `presumeLoginMethod` or `forceLoginMethod` depends on whether you allow the user to go "Back" and choose a different method, or if this is the only option available to them. Though I suppose the "force" could be achieved simply by disabling all but one of the login methods.

Reply:
@bradenmacdonald

I'm not sure how I feel about this suggestion-- for two reasons. The first is that there's a rendering effect I'm concerned about here. The `forceWebLogin` method is there not only to bring the user to the web authentication page, but to make sure that, on initial load, the sign in page isn't drawn. There needs to be some way to signal to the view not to render the button or text if we're going to be opening the Safari browser to perform the login. Otherwise you get this visual glitch where you launch the app, see the sign-in page, and then the browser opens, which can be alarming and confusing-- I know that if I saw that as a user, I'd wonder if I pressed something I shouldn't or something happened I didn't want.

What you suggest could notionally be compatible with that first reason, but it would still require special casing the web login because, presumably, any other login method would have its interfaces written into the Swift view, and thus would need to be hidden if we're presuming the web login.

The second reason I'm not sure about your suggestion is that I'm not sure any other login methods, other than direct login, are actually viable outside of a web view, or wouldn't otherwise require some other complex changes to the UI. For example if you were to sign in with Google, you'd still need to visit the LMS login page to do the same dance we're doing for the client with Auth0. Same with any SAML provider, as far as I can tell-- the existing login view assumes that the user will be logging into the LMS using a post to Django directly, but nearly all external authentication methods will require visiting some intermediary web page and getting redirected back, requiring a Safari window, or else some tighter, complex iOS integration I'm unaware of.

For reference, I initially attempted to do the integration following Auth0's guide on integrating with iOS. However I realized as I was working with it that while I could authenticate with Auth0, the tokens it would provide me only worked with their servers directly. I couldn't get a token for the LMS without visiting the LMS and getting redirected from it to a login page provided by Auth0. There might be some other way to do it but it escapes me for now, and I suspect this is the way it's intended to be done. After all, all the MFEs at present just redirect to the LMS for login and then redirect you back, rather than presenting a streamlined login MFE.